src/v/cluster/partition_balancer_planner.cc

```diff
@@ -0,0 +1,422 @@
+ */
+
+#include "cluster/partition_balancer_planner.h"
+
+#include "cluster/partition_balancer_types.h"
+#include "cluster/scheduling/constraints.h"
+#include "cluster/scheduling/types.h"
+
+#include <optional>
+
+namespace cluster {
+
+partition_balancer_planner::partition_balancer_planner(
+  planner_config config,
+  topic_table& topic_table,
+  partition_allocator& partition_allocator)
+  : _config(config)
+  , _topic_table(topic_table)
+  , _partition_allocator(partition_allocator) {}
+
+void partition_balancer_planner::
+  init_ntp_sizes_and_node_disk_reports_from_health_report(
+    const cluster_health_report& health_report,
+    reallocation_request_state& rrs) {
+    for (const auto& node_report : health_report.node_reports) {
+        uint64_t total = 0;
+        uint64_t free = 0;
+        for (const auto& disk : node_report.local_state.disks) {
+            total += disk.total;
+            free += disk.free;
+        }
+        rrs.node_disk_reports.insert(
+          {node_report.id, node_disk_space(node_report.id, free, total)});
+        for (const auto& tp_ns : node_report.topics) {
+            for (const auto& partition : tp_ns.partitions) {
+                rrs.ntp_sizes[model::ntp(
+                  tp_ns.tp_ns.ns, tp_ns.tp_ns.tp, partition.id)]
+                  = partition.size_bytes;
+            }
+        }
+    }
+    return;
+}
+
+void partition_balancer_planner::
+  calculate_nodes_with_disk_constraints_violation(
+    reallocation_request_state& rrs) {
+    for (const auto& n : rrs.node_disk_reports) {
+        if (
+          n.second.free_space_rate < double(1) - _config.max_disk_usage_ratio) {
+            rrs.nodes_with_disk_constraints_violation.insert(n.first);
+            rrs.full_nodes_disk_sizes.insert(n.second);
+        }
+    }
+}
+
+void partition_balancer_planner::calculate_unavailable_nodes(
+  const std::vector<raft::follower_metrics>& follower_metrics,
+  reallocation_request_state& rrs) {
+    const auto now = raft::clock_type::now();
+    for (const auto& follower : follower_metrics) {
+        if (
+          now - follower.last_heartbeat
+          > _config.node_availability_timeout_sec) {
+            rrs.unavailable_nodes.insert(follower.id);
+        }
+    }
+}
+
+std::vector<model::broker_shard>
+partition_balancer_planner::get_stable_replicas(
+  const std::vector<model::broker_shard>& replicas,
+  const reallocation_request_state& rrs,
+  size_t full_nodes_leave_amount) {
+    std::vector<model::broker_shard> stable_replicas;
+    absl::flat_hash_map<model::node_id, uint32_t>
+      replicas_breaking_disk_constraints;
+
+    // add stable nodes
+    for (const auto& r : replicas) {
+        if (!rrs.unavailable_nodes.contains(r.node_id)) {
+            if (!rrs.nodes_with_disk_constraints_violation.contains(
+                  r.node_id)) {
+                stable_replicas.push_back(
+                  model::broker_shard{r.node_id, r.shard});
```

Comment:
nit: can we just use `stable_replicas.push_back(r)` ?

Reply:
fixed